### PR TITLE
fix(nm): Settings leak on connection activation failure

### PR DESCRIPTION
**Pull request by GregoryIvo**
This PR modifies the [GetAppliedConnection()](https://networkmanager.dev/docs/api/latest/gdbus-org.freedesktop.NetworkManager.Device.html#gdbus-method-org-freedesktop-NetworkManager-Device.GetAppliedConnection) method.

Old behaviour: would find an associated connection and update an existing one. If it doesn’t it creates a new one.

New Behaviour: check for associated connections, then check for non-active associated connections. If more than one associated connection is found it will take the first and delete the rest. If none is found, create a new one.

> **Note**: We are using the Conventional Commits convention for our pull request titles. Please take a look at the [PR title format document](https://github.com/eclipse/kura/blob/develop/CONTRIBUTING.md#submitting-the-changes) for the supported [types](https://github.com/eclipse/kura/blob/develop/CONTRIBUTING.md#type) and [scopes](https://github.com/eclipse/kura/blob/develop/CONTRIBUTING.md#scope).

Brief description of the PR. [e.g. Added `null` check on `object` to avoid `NullPointerException`]

**Related Issue:** This PR fixes/closes {issue number}

**Description of the solution adopted:** A more detailed description of the changes made to solve/close one or more issues. If the PR is simple and easy to understand this section can be skipped.

**Screenshots:** If applicable, add screenshots to help explain your solution

**Any side note on the changes made:** Description of any other change that has been made, which is not directly linked to the issue resolution [e.g. Code clean up/Sonar issue resolution]

### CODE REVIEW

**Comment by GregoryIvo** [comment]
@mattdibi 

is getInterfaceIds the new valid replacement for this API?
```
[ERROR] Failed to execute goal org.eclipse.tycho:tycho-compiler-plugin:1.7.0:compile (default-compile) on project org.eclipse.kura.nm: Compilation failure: Compilation failure: 
[ERROR] /home/jenkins/agent/workspace/multibranch_PR-4477/kura/kura/org.eclipse.kura.nm/src/main/java/org/eclipse/kura/nm/NMDbusConnector.java:[548] 
[ERROR] 	return getAvaliableConnection(dev);
[ERROR] 	^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
[ERROR] Unreachable code
[ERROR] -> [Help 1]
[ERROR] 
[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR] 
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MojoFailureException
[ERROR] 
[ERROR] After correcting the problems, you can resume the build with the command
[ERROR]   mvn <args> -rf :org.eclipse.kura.nm
```

**Comment by MMaiero** [comment]
```
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-compiler-plugin:3.10.1:testCompile (compiletests) on project org.eclipse.kura.nm.test: Compilation failure
[ERROR] /home/jenkins/agent/workspace/multibranch_PR-4477/kura/kura/test/org.eclipse.kura.nm.test/src/test/java/org/eclipse/kura/nm/NMDbusConnectorTest.java:[107,5] cannot find symbol
[ERROR]   symbol:   class NetworkService
[ERROR]   location: class org.eclipse.kura.nm.NMDbusConnectorTest
[ERROR] -> [Help 1]
[ERROR] 
[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR] 
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MojoFailureException
[ERROR] 
[ERROR] After correcting the problems, you can resume the build with the command
[ERROR]   mvn <args> -rf :org.eclipse.kura.nm.test
```

File: kura/org.eclipse.kura.nm/src/main/java/org/eclipse/kura/nm/NMDbusConnector.java [hunk 1]
```diff
@@ -71,6 +71,7 @@ public class NMDbusConnector {
     private static final String NM_DEVICE_PROPERTY_DEVICETYPE = "DeviceType";
     private static final String NM_DEVICE_PROPERTY_STATE = "State";
     private static final String NM_DEVICE_PROPERTY_IP4CONFIG = "Ip4Config";
+    private static final String NM_DEVICE_PROPERTY_CONFIGURATION = "connection";
```
Comment (mattdibi) [inline kura/org.eclipse.kura.nm/src/main/java/org/eclipse/kura/nm/NMDbusConnector.java]: This is not a device property and is not called "configuration". This doesn't follow the already established naming convention.

Comment (GregoryIvo) [reply kura/org.eclipse.kura.nm/src/main/java/org/eclipse/kura/nm/NMDbusConnector.java]: I think this got changed during the cherry picking mess. 

Is ..._connection okay?

**Comment by MMaiero** [comment]
@GregoryIvo Are we done with the changes for this PR?

**Comment by GregoryIvo** [comment]
> @GregoryIvo Are we done with the changes for this PR?

With these last few changes we are. Just need a last review from Mattia.

**Comment by MMaiero** [comment]
Thanks. Can you have a look at the conflicting file?

**Comment by MMaiero** [comment]
@mattdibi It should be ready now for review

**Comment by mattdibi** [comment]
I'm on it

**Review by mattdibi — CHANGES_REQUESTED** [review]
As per our discussion:
- The housekeeping won't run if there's an active connection available.
- `getFirstAvailableConnection` and `cleanAvailableConnection` show a lot of code duplication
- I don't like the fact that `getAssociatedConnection` have side-effects
- Code can be simplified
- Add more test cases

Some of these issues were due to how I (wrongly) initially designed the code. We already addressed them in my commits. Still some changes pending

File: kura/test/org.eclipse.kura.nm.test/src/test/java/org/eclipse/kura/nm/NMDbusConnectorTest.java [hunk 2]
```diff
@@ -858,6 +890,48 @@ private void givenMockedDevice(String deviceId, String interfaceId, NMDeviceType
...
+        
+        Map<String, Map<String, Variant<?>>> connectionSettings = new HashMap<>();
+
+        Map<String, Variant<?>> variantConfig = new HashMap<>();
+
+        variantConfig.put("id", new Variant<>(connectionId));
+        variantConfig.put("uuid", new Variant<>(connectionUuid));
+        connectionSettings.put("connection", variantConfig);
+
+        Connection mockConnection = mock(Connection.class);
```
Comment (mattdibi) [inline kura/test/org.eclipse.kura.nm.test/src/test/java/org/eclipse/kura/nm/NMDbusConnectorTest.java]: Watch out! You're hiding the `mockConnection` variable at line 112 here.

File: kura/test/org.eclipse.kura.nm.test/src/test/java/org/eclipse/kura/nm/NMDbusConnectorTest.java [hunk 3]
```diff
@@ -1184,6 +1258,14 @@ private void thenNetInterfaceTypeIs(NetworkInterfaceType type) {
     public void thenConfigurationEnforcementIsActive(boolean expectedValue) {
         assertEquals(expectedValue, this.instanceNMDbusConnector.configurationEnforcementIsActive());
     }
+    
+    private void thenConnectionIsDeleted(String path) {
+        verify(mockedConnections.get(path)).Delete();
```
Comment (mattdibi) [inline kura/test/org.eclipse.kura.nm.test/src/test/java/org/eclipse/kura/nm/NMDbusConnectorTest.java]: Be specific. You expect only *one* call. If there's more than one there's something wrong.

```suggestion
        verify(mockedConnections.get(path), times(1)).Delete();
```

File: kura/test/org.eclipse.kura.nm.test/src/test/java/org/eclipse/kura/nm/NMDbusConnectorTest.java (hunk 2 above)
Comment (GregoryIvo) [reply kura/test/org.eclipse.kura.nm.test/src/test/java/org/eclipse/kura/nm/NMDbusConnectorTest.java:925]: I don't see what the issue is with this?

**Comment by GregoryIvo** [comment]
merging because re-base was incompatible

**Review by mattdibi — APPROVED** [review]

**Comment by MMaiero** [comment]
Forcing merge. The feature will be further tested in the upcoming RC1 build

**Pull request merged**